### ISSUE TITLE
Enable tutorial.error.checker to provide checks on error

### ISSUE THREAD

**Issue by garrettgman** [rstudio/learnr#356]
If a student submits code that throws an error, learnr exits the checking routine and displays the error message. This short circuit prevents the checking code from telling the student anything useful about what might have gone wrong. In short, the student is left to decipher the error message on his or her own.

This is necessary when the checking code relies on the result of the student code (if it throws an error, there is no result). But it is a missed opportunity when the checking code parses the unevaluated student code, as `gradethis::grade_code()` does. Here is an example, `grade_code()` would return `"I expected mtcars where you wrote mt"` but learnr displays the error message. (code appended at bottom)

<img width="445" alt="Screen Shot 2020-04-23 at 2 58 41 PM" src="https://user-images.githubusercontent.com/85014/80138629-121f5d00-8573-11ea-8fa1-57d583f14460.png">

To fix this, I suggest that we:
1. Add a `tutorials.error.checker` option that can be set to a grading function to run in the event that the student code throws an error. This option can be set in the same manner as the `tutorials.exercise.checker` option. It should be set to a function that can grade the student code without evaluating it, similar to `gradethis::grade_code()`.
<img width="300" alt="Screen Shot 2020-04-23 at 3 19 10 PM" src="https://user-images.githubusercontent.com/85014/80235936-5ae51d80-8628-11ea-8dff-69bd0556effe.png">

 
2. Display the result of the error checking function beneath the error message. Here is a mock up. 
<img width="445" alt="error-mockup 001" src="https://user-images.githubusercontent.com/85014/80236215-d34bde80-8628-11ea-8096-ea4bccb5a8d7.png">

Code to recreate existing example above
****
````
---
title: "Example"
output: learnr::tutorial
runtime: shiny_prerendered
---

```{r setup, include=FALSE}
library(learnr)
library(gradethis)

tutorial_options(exercise.checker = gradethis::grade_learnr)
knitr::opts_chunk$set(out.width = "100%")
```

## Welcome

### `mtcars`

To see the `mtcars` data frame, type `mtcars` in the code block below and click "Submit Answer".

```{r ex, exercise = TRUE}

```

```{r ex-solution}
mtcars
```

```{r ex-check}
grade_code("Good job!")
```
````

**Comment by garrettgman** [comment]
At the moment, this is the chief reason I'm delaying wiring up an autograder to the primers.

**Comment by schloerke** [comment]
Plan of action:

* Have learnr determine if it is a syntax error. Allow for global option to set message of syntax error.
* If an evaluation error, call `-check` chunk with the result as the error object.  The grading code should handle the error object.

**Comment by dtkaplan** [comment]
I endorse Garrett's point about enabling the grading function to handle parsing errors and evaluation errors. Let me describe a strategy that I think would add a lot of flexibility and power to  people writing error-checking packages. You can leave the current `exercise=TRUE` system and all its fellow `-hint`, `-solution`, `-check` etc. chunks as they are. 

Create another type of user-input widget. Let's call in it a `submission=TRUE` object. It could look identical to the `exercise` widget but doesn't have to. When the user presses "Run code" or "hint" or "submit", the learnr system would set up a data structure and hand that off to an author-specified `handle_submission()` function. The data structure should include:

1. The content that the student put in the text field.
2. Which button the student pressed: "hint" or "submit" or "run code."
3. A list with the content  of all chunks in the document, along with the chunk label and a list of chunk parameters.
4. A `uiOutput` object  associated with the chunk and located just below it. (And, I'd suggest another one  just above it to replace the current 'hint' display. But leave that aside for the moment.)

Leave it to the exercise checker to evaluate any setup  code from (3) and the student submission code (1).

Learnr should provide functions that allows the checker system to assign specific shiny output widgets to the `uiOutput` associated with the submission chunk, and write to those widgets in the usual shiny way.

Learnr should not evaluate the student submission directly.  Leave that to the checker code, with learnr providing a safe sandbox in which that checker code can operate.

My experience in writing `checkr`, which  was possibly the first learnr checking  package,  was that I don't need learnr to be providing  environments and values, etc. I can do all that more effectively in the checking code.

Maybe a short way to describe what I'd like ...

- a way to display a module of shiny inputs and outputs
- a function that will return the contents of all chunks in the document, along with  labels, etc.  
- a safe place to evaluate the server components of the shiny module
- a way to trigger a learnr event so that the results of the checking can become part of the stream of logging events from the learnr document.

**Comment by garrettgman** [comment]
I think of the `-check` chunk (and `grade_result()` ) as the grading
equivalent of unit tests. I'm trying to anticipate ahead of time all of the
ways the student can fail, and to provide targeted guidance for each
failure mode. Here, `cars` is not the right answer, but it is a failure
mode that I want to screen for, because the two data set names are easy to
confuse in your head (BTW this is a made up example).

Most "unit tests" will need to test the result of the student code. But the
final "unit test" would always be "was the result an error? If so return
the output of `grade_code()`." The tricky thing is to make sure that _all_
of the tests get applied to the student work in every case.

TL;DR - I want them to get mtcars because that is the solution, but I want
to include a check for the failure mode where they use cars.

**Comment by cpsievert** [comment]
> But the final "unit test" would always be "was the result an error? If so return the output of `grade_code()`." 

Sounds like this implementation of `exercise.error.checker` will do want you want:

```r
function(solution_code = NULL, check_code, ...) {
   if (is.null(solution_code)) {
     return(NULL)
   }
  grade_learnr(solution_code = solution_code, check_code = "grade_code()", ...)
}
```

This'll make sure `grade_code()` is run when an evaluation error occurs and a solution is provided.  This is different from the current implementation of `grade_learnr_error` (maybe we should change it?), which currently runs the `-check` chunk code on the evaluation error.

```r
function(solution_code = NULL, check_code, ...) {
   if (is.null(solution_code)) {
     return(NULL)
   }
  grade_learnr(solution_code = solution_code, check_code = check_code, ...)
}
```
